File: src/main/mods/streaming/infrastructure/fallback-turn-server-provider.ts
import type { StreamingTurnServerConfig } from '../../../../shared/rpc/streaming'
import type { StreamingTargetType } from '../domain/types'

const HOME_FALLBACK_TURN_SERVER_URL = [
  'https://',
  'x',
  'streaming-support.pages.dev/server.json'
].join('')

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function normalizeTurnServerConfig(value: unknown): StreamingTurnServerConfig | null {
  if (!isRecord(value)) {
    return null
  }

  const url = typeof value.url === 'string' ? value.url.trim() : ''
  const username = typeof value.username === 'string' ? value.username.trim() : ''
  const credential = typeof value.credential === 'string' ? value.credential.trim() : ''
  if (url === '' || username === '' || credential === '') {
    return null
  }

  return {
    url,
    username,
    credential
  }
}

/**
 * fallback TURN 提供器
 * - 将兜底 TURN 的发现逻辑放回主进程，统一缓存与替换入口
 */
export class FallbackTurnServerProvider {
  private homeTurnServer: StreamingTurnServerConfig | null | undefined
  private homeTurnServerPromise: Promise<StreamingTurnServerConfig | null> | null = null

  async getByTargetType(type: StreamingTargetType): Promise<StreamingTurnServerConfig | null> {
    if (type !== 'home') {
      return null
    }

    if (this.homeTurnServer !== undefined) {
      return this.homeTurnServer === null ? null : { ...this.homeTurnServer }
    }

    if (this.homeTurnServerPromise !== null) {
      const result = await this.homeTurnServerPromise
      return result === null ? null : { ...result }
    }

    this.homeTurnServerPromise = this.fetchHomeTurnServer()
    try {
      const result = await this.homeTurnServerPromise
      this.homeTurnServer = result
      return result === null ? null : { ...result }
    } finally {
      this.homeTurnServerPromise = null
    }
  }

  private async fetchHomeTurnServer(): Promise<StreamingTurnServerConfig | null> {
    try {
      const response = await fetch(HOME_FALLBACK_TURN_SERVER_URL, {
        signal: AbortSignal.timeout(10_000)
      })
      if (!response.ok) {
        return null
      }

      return normalizeTurnServerConfig(await response.json())
    } catch (error) {
      console.warn('[Streaming] load fallback turn server failed:', error)
      return null
    }
  }
}

File: src-tauri/src/mods/streaming/fallback_turn_server_provider.rs
use crate::mods::streaming::types::StreamingTurnServerConfig;
use reqwest::Client;

const HOME_FALLBACK_TURN_SERVER_URL: &str = "https://xstreaming-support.pages.dev/server.json";

pub struct FallbackTurnServerProvider {
    home_turn_server: Option<Option<StreamingTurnServerConfig>>,
    client: Client,
}

impl FallbackTurnServerProvider {
    pub fn new() -> Self {
        Self {
            home_turn_server: None,
            client: Client::new(),
        }
    }

    pub async fn get_by_target_type(
        &mut self,
        target_type: &str,
    ) -> Result<Option<StreamingTurnServerConfig>, String> {
        if target_type != "home" {
            return Ok(None);
        }

        if let Some(value) = &self.home_turn_server {
            return Ok(value.clone());
        }

        let value = self.fetch_home_turn_server().await;
        self.home_turn_server = Some(value.clone());
        Ok(value)
    }

    async fn fetch_home_turn_server(&self) -> Option<StreamingTurnServerConfig> {
        let response = match self
            .client
            .get(HOME_FALLBACK_TURN_SERVER_URL)
            .timeout(std::time::Duration::from_secs(10))
            .send()
            .await
        {
            Ok(response) => response,
            Err(_) => return None,
        };

        if !response.status().is_success() {
            return None;
        }

        let value = response.json::<serde_json::Value>().await.ok()?;
        let url = value.get("url").and_then(|item| item.as_str())?.trim();
        let username = value.get("username").and_then(|item| item.as_str())?.trim();
        let credential = value
            .get("credential")
            .and_then(|item| item.as_str())?
            .trim();

        if url.is_empty() || username.is_empty() || credential.is_empty() {
            return None;
        }

        Some(StreamingTurnServerConfig {
            url: url.to_string(),
            username: username.to_string(),
            credential: credential.to_string(),
        })
    }
}
